README.md
Description
===========
Installs and configures Jenkins CI server & node slaves.  Resource providers to support automation via jenkins-cli, including job create/update.

Requirements
============

Chef
----

* Chef version 0.9.10 or higher

Platform
--------

* 'default' - Server installation - currently supports Red Hat/CentOS 5.x and Ubuntu 8.x/9.x/10.x

* 'node_ssh' - Any platform that is running sshd.

* 'node_jnlp' - Unix platforms. (depends on runit recipe)

* 'node_windows' - Windows platforms only.  Depends on .NET Framework, which can be installed with the windows::dotnetfx recipe.

Java
----

Jenkins requires Java 1.5 or higher, which can be installed via the Opscode java cookbook or windows::java recipe.

Jenkins node authentication
---------------------------

If your Jenkins instance requires authentication, you'll either need to embed user:pass in the jenkins.server.url or issue a jenkins-cli.jar login command prior to using the jenkins::node_* recipes.  For example, define a role like so:

  name "jenkins_ssh_node"
  description "cli login & register ssh slave with Jenkins"
  run_list %w(vmw::jenkins_login jenkins::node_ssh)

Where the jenkins_login recipe is simply:

  jenkins_cli "login --username #{node[:jenkins][:username]} --password #{node[:jenkins][:password]}"

Attributes
==========

* jenkins[:mirror] - Base URL for downloading Jenkins (server)
* jenkins[:java_home] - Java install path, used for for cli commands
* jenkins[:server][:home] - JENKINS_HOME directory
* jenkins[:server][:user] - User the Jenkins server runs as
* jenkins[:server][:group] - Jenkins user primary group
* jenkins[:server][:port] - TCP listen port for the Jenkins server
* jenkins[:server][:url] - Base URL of the Jenkins server
* jenkins[:server][:plugins] - Download the latest version of plugins in this list, bypassing update center
* jenkins[:node][:name] - Name of the node within Jenkins
* jenkins[:node][:description] - Jenkins node description
* jenkins[:node][:executors] - Number of node executors
* jenkins[:node][:home] - Home directory ("Remote FS root") of the node
* jenkins[:node][:labels] - Node labels
* jenkins[:node][:mode] - Node usage mode, "normal" or "exclusive" (tied jobs only)
* jenkins[:node][:launcher] - Node launch method, "jnlp", "ssh" or "command"
* jenkins[:node][:availability] - "always" keeps node on-line, "demand" off-lines when idle
* jenkins[:node][:in_demand_delay] - number of minutes for which jobs must be waiting in the queue before attempting to launch this slave.
* jenkins[:node][:idle_delay] - number of minutes that this slave must remain idle before taking it off-line.
* jenkins[:node][:env] - "Node Properties" -> "Environment Variables"
* jenkins[:node][:user] - user the slave runs as
* jenkins[:node][:ssh_host] - Hostname or IP Jenkins should connect to when launching an SSH slave
* jenkins[:node][:ssh_port] - SSH slave port
* jenkins[:node][:ssh_user] - SSH slave user name (only required if jenkins server and slave user is different)
* jenkins[:node][:ssh_pass] - SSH slave password (not required when server is installed via default recipe)
* jenkins[:node][:ssh_private_key] - jenkins master defaults to: `~/.ssh/id_rsa` (created by the default recipe)
* jenkins[:node][:jvm_options] - SSH slave JVM options
* jenkins[:iptables_allow] - if iptables is enabled, add a rule passing 'jenkins[:server][:port]'
* jenkins[:http_proxy][:variant] - use `nginx` or `apache2` to proxy traffic to jenkins backend (`nil` by default)
* jenkins[:http_proxy][:www_redirect] - add a redirect rule for 'www.*' URL requests ("disable" by default)
* jenkins[:http_proxy][:listen_ports] - list of HTTP ports for the HTTP proxy to listen on ([80] by default)
* jenkins[:http_proxy][:host_name] - primary vhost name for the HTTP proxy to respond to (`node[:fqdn]` by default)
* jenkins[:http_proxy][:host_aliases] - optional list of other host aliases to respond to (empty by default)
* jenkins[:http_proxy][:client_max_body_size] - max client upload size ("1024m" by default, nginx only)

Usage
=====

'default' recipe
----------------

Installs a Jenkins CI server using the http://jenkins-ci.org/redhat RPM.  The recipe also generates an ssh private key and stores the ssh public key in the node 'jenkins[:pubkey]' attribute for use by the node recipes.

'node_ssh' recipe
-----------------

Creates the user and group for the Jenkins slave to run as and sets `.ssh/authorized_keys` to the 'jenkins[:pubkey]' attribute.  The 'jenkins-cli.jar'[1] is downloaded from the Jenkins server and used to manage the nodes via the 'groovy'[2] cli command.  Jenkins is configured to launch a slave agent on the node using its SSH slave plugin[3].

[1] http://wiki.jenkins-ci.org/display/JENKINS/Jenkins+CLI
[2] http://wiki.jenkins-ci.org/display/JENKINS/Jenkins+Script+Console
[3] http://wiki.jenkins-ci.org/display/JENKINS/SSH+Slaves+plugin

'node_jnlp' recipe
------------------

Creates the user and group for the Jenkins slave to run as and '/jnlpJars/slave.jar' is downloaded from the Jenkins server.  Depends on runit_service from the runit cookbook.

'node_windows' recipe
---------------------

Creates the home directory for the node slave and sets 'JENKINS_HOME' and 'JENKINS_URL' system environment variables.  The 'winsw'[1] Windows service wrapper will be downloaded and installed, along with generating `jenkins-slave.xml` from a template.  Jenkins is configured with the node as a 'jnlp'[2] slave and '/jnlpJars/slave.jar' is downloaded from the Jenkins server.  The 'jenkinsslave' service will be started the first time the recipe is run or if the service is not running.  The 'jenkinsslave' service will be restarted if '/jnlpJars/slave.jar' has changed.  The end results is functionally the same had you chosen the option to "Let Jenkins control this slave as a Windows service"[3].

[1] http://weblogs.java.net/blog/2008/09/29/winsw-windows-service-wrapper-less-restrictive-license
[2] http://wiki.jenkins-ci.org/display/JENKINS/Distributed+builds
[3] http://wiki.jenkins-ci.org/display/JENKINS/Installing+Jenkins+as+a+Windows+service

'proxy_nginx' recipe
--------------------

Uses the nginx::source recipe from the nginx cookbook to install an HTTP frontend proxy. To automatically activate this recipe set the `node[:jenkins][:http_proxy][:variant]` to `nginx`.

'proxy_apache2' recipe
----------------------

Uses the apache2 recipe from the apache2 cookbook to install an HTTP frontend proxy. To automatically activate this recipe set the `node[:jenkins][:http_proxy][:variant]` to `apache2`.

'builds' data bag
-----------------

Create a data bag called "builds" with the following attributes to specify a Jenkins job that creates an EC2 instance configured via Chef for deployment of the project's assets for testing:
    
    "id" = Name of the data bag item
    "repository" = SCM URL repository (i.e. git@github.com:you/your_project)
    "branch" = Branch to check out of SCM
    "component" = Name of the component/project that you are building
    "build_on_change" = true/false, whether to invoke a build after any new job is created or whether the job config changes
    "groupId" = maven group id (optional)
    "artifactId" = maven artifact id (optional)
    "goal" = maven goal to run in the jenkins job
    "roles" = a comma-seperated list of roles to run on the node (e.g. role[my_app_server],role[my_test_support])
    "ami" = the AMI ID to provision the EC2 instance with
    "server_size" = The EC2 server type to use (i.e. m1.small)
    "ec2_ssh_key" = EC2's ssh server key to use
    "ec2_security_group" = EC2 security group to configure the server's firewall with
    "chef_bootstrap" = Chef bootstrap (i.e. ubuntu10.04-gems)
    "ami_user" = The default user provisioned by the AMI for remote SSH login
    "ec2_ssh_key_file" = (Yet another) SSH key access, .pem file.
    "ec2_region" : EC2 region (i.e. us-east-1)


'jenkins_cli' resource provider
-------------------------------

This resource can be used to execute the Jenkins cli from your recipes.  For example, install plugins via update center and restart Jenkins:

    %w(git URLSCM build-publisher).each do |plugin|
      jenkins_cli "install-plugin #{plugin}"
      jenkins_cli "safe-restart"
    end

'jenkins_node' resource provider
--------------------------------

This resource can be used to configure nodes as the 'node_ssh' and 'node_windows' recipes do or "Launch slave via execution of command on the Master".

    jenkins_node node[:fqdn] do
      description  "My node for things, stuff and whatnot"
      executors    5
      remote_fs    "/var/jenkins"
      launcher     "command"
      command      "ssh -i my_key #{node[:fqdn]} java -jar #{remote_fs}/slave.jar"
      env          "ANT_HOME" => "/usr/local/ant", "M2_REPO" => "/dev/null"
    end

'jenkins_job' resource provider
-------------------------------

This resource manages jenkins jobs, supporting the following actions:

   :create, :update, :delete, :build, :disable, :enable

The 'create' and 'update' actions require a jenkins job config.xml.  Example:

    git_branch = 'master'
    job_name = "sigar-#{branch}-#{node[:os]}-#{node[:kernel][:machine]}"

    job_config = File.join(node[:jenkins][:node][:home], "#{job_name}-config.xml")

    jenkins_job job_name do
      action :nothing
      config job_config
    end

    template job_config do
      source "sigar-jenkins-config.xml"
      variables :job_name => job_name, :branch => git_branch, :node => node[:fqdn]
      notifies :update, resources(:jenkins_job => job_name), :immediately
      notifies :build, resources(:jenkins_job => job_name), :immediately
    end

'manage_node' library
---------------------

The script to generate groovy that manages a node can be used standalone.  For example:

    % ruby manage_node.rb name slave-hostname remote_fs /home/jenkins ... | java -jar jenkins-cli.jar -s http://jenkins:8080/ groovy =

Issues
======

* CLI authentication - http://issues.jenkins-ci.org/browse/JENKINS-3796

* CLI *-node commands fail with "No argument is allowed: nameofslave" - http://issues.jenkins-ci.org/browse/JENKINS-5973

License & Author
================

This is a downstream fork of Doug MacEachern's Hudson cookbook (https://github.com/dougm/site-cookbooks) and therefore deserves all the glory.

Author:: Doug MacEachern (<dougm@vmware.com>)

Contributor:: AJ Christensen <aj@junglist.gen.nz>
Contributor:: Fletcher Nichol <fnichol@nichol.ca>
Contributor:: Roman Kamyk <rkj@go2.pl>
Contributor:: Darko Fabijan <darko@renderedtext.com>

Copyright:: 2010, VMware, Inc

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG

Fix a few issues and add features:
 - create data bag for builds that allows one to manage a Jenkins job
   that creates an EC2 instance configure via a set of Chef roles for
   deployment and testing
 - add job creation/updates to the default recipe via data bag mechanism
 - add "update-job" Jenkins CLI capability to the job resource
 - add HTTP Basic Auth support to nginx proxy
 - fix issues relating to ubuntu support including:
   - make the jenkins service start at the end of installation
   - update the apt packages after adding the jenkins debian repository
   - supply a correct JAVA_HOME to the jenkins upstart configuration file
     /etc/init/jenkins.conf, as well as support changing the port that
     jenkins listens on
   - configure jenkins for the location of the maven installation if the
     maven recipe is installed
 - add nginx as a dependency
 - correct README on attributes for proxy variants
 - configure for github access via ssh
 - bind to localhost if using any proxy




recipes/proxy_nginx.rb
#
# Cookbook Name:: jenkins
# Recipe:: proxy_nginx
#
# Author:: Fletcher Nichol <fnichol@nichol.ca>
#
# Copyright 2011, Fletcher Nichol
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe "nginx"

if node[:jenkins][:http_proxy][:www_redirect] == "enable"
  www_redirect = true
else
  www_redirect = false
end

host_name = node[:jenkins][:http_proxy][:host_name] || node[:fqdn]

template "#{node[:nginx][:dir]}/htpasswd" do
  variables( :username => node[:jenkins][:http_proxy][:basic_auth_username],
             :password => node[:jenkins][:http_proxy][:basic_auth_password])
  owner       node[:nginx][:user]
  group       'root'
  mode 0600
end

template "#{node[:nginx][:dir]}/sites-available/jenkins.conf" do
  source      "nginx_jenkins.conf.erb"
  owner       'root'
  group       'root'
  mode        '0644'
  variables(
    :host_name        => host_name,
    :host_aliases     => node[:jenkins][:http_proxy][:host_aliases],
    :listen_ports     => node[:jenkins][:http_proxy][:listen_ports],
    :www_redirect     => www_redirect,
    :max_upload_size  => node[:jenkins][:http_proxy][:client_max_body_size]
  )

  if File.exists?("#{node[:nginx][:dir]}/sites-enabled/jenkins.conf")
    notifies  :restart, 'service[nginx]'
  end
end

nginx_site "jenkins.conf" do
  if node[:jenkins][:http_proxy][:variant] == "nginx"
    enable true
  else
    enable false
  end
end


recipes/default.rb
#
# Cookbook Name:: jenkins
# Based on hudson
# Recipe:: default
#
# Author:: AJ Christensen <aj@junglist.gen.nz>
# Author:: Doug MacEachern <dougm@vmware.com>
# Author:: Fletcher Nichol <fnichol@nichol.ca>
#
# Copyright 2010, VMware, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

pkey = "#{node[:jenkins][:server][:home]}/.ssh/id_rsa"
tmp = "/tmp"

user node[:jenkins][:server][:user] do
  home node[:jenkins][:server][:home]
end

directory node[:jenkins][:server][:home] do
  recursive true
  owner node[:jenkins][:server][:user]
  group node[:jenkins][:server][:group]
end

directory "#{node[:jenkins][:server][:home]}/.ssh" do
  mode 0700
  owner node[:jenkins][:server][:user]
  group node[:jenkins][:server][:group]
end

#cookbook_file "#{pkey}" do
  #source "id_rsa"
  #mode 0600
  #owner node[:jenkins][:server][:user]
  #group node[:jenkins][:server][:group]
#end

#cookbook_file "#{pkey}.pub" do
  #source "id_rsa.pub"
  #mode 0644
  #owner node[:jenkins][:server][:user]
  #group node[:jenkins][:server][:group]
#end

cookbook_file "#{node[:jenkins][:server][:home]}/.ssh/known_hosts" do
  source "github_keys.pub"
  mode 0644
  owner node[:jenkins][:server][:user]
  group node[:jenkins][:server][:group]
end


execute "ssh-keygen -f #{pkey} -N ''" do
  user  node[:jenkins][:server][:user]
  group node[:jenkins][:server][:group]
  not_if { File.exists?(pkey) }
end

ruby_block "store jenkins ssh pubkey" do
  block do
    node.set[:jenkins][:server][:pubkey] = File.open("#{pkey}.pub") { |f| f.gets }
  end
end

template "#{node[:jenkins][:server][:home]}/.gitconfig" do
  source "dot_gitconfig.erb"
  mode 0664
  owner node[:jenkins][:server][:user]
  group node[:jenkins][:server][:group]
end

directory "#{node[:jenkins][:server][:home]}/plugins" do
  owner node[:jenkins][:server][:user]
  group node[:jenkins][:server][:group]
  only_if { node[:jenkins][:server][:plugins].size > 0 }
end

node[:jenkins][:server][:plugins].each do |name|
  remote_file "#{node[:jenkins][:server][:home]}/plugins/#{name}.hpi" do
    source "#{node[:jenkins][:mirror]}/plugins/#{name}/latest/#{name}.hpi"
    backup false
    owner node[:jenkins][:server][:user]
    group node[:jenkins][:server][:group]
    action :create_if_missing
  end
end

case node.platform
when "ubuntu", "debian"
  include_recipe "apt"
  include_recipe "java"

  pid_file = "/var/run/jenkins/jenkins.pid"
  install_starts_service = false

  apt_repository "jenkins" do
    uri "#{node.jenkins.package_url}/debian"
    components %w[binary/]
    key "http://pkg.jenkins-ci.org/debian/jenkins-ci.org.key"
    #action :add
  end
  cookbook_file "/etc/apt/sources.list.d/jenkins.list" do
    mode        '0644'
    source "jenkins.list"
    notifies :run, resources(:execute => "apt-get update"), :immediately
  end

when "centos", "redhat"
  include_recipe "yum"

  pid_file = "/var/run/jenkins.pid"
  install_starts_service = false

  yum_key "jenkins" do
    url "#{node.jenkins.package_url}/redhat/jenkins-ci.org.key"
    action :add
  end

  yum_repository "jenkins" do
    description "repository for jenkins"
    url "#{node.jenkins.package_url}/redhat/"
    key "jenkins"
    action :add
  end
end

#"jenkins stop" may (likely) exit before the process is actually dead
#so we sleep until nothing is listening on jenkins.server.port (according to netstat)
ruby_block "netstat" do
  block do
    10.times do
      if IO.popen("netstat -lnt").entries.select { |entry|
          entry.split[3] =~ /:#{node[:jenkins][:server][:port]}$/
        }.size == 0
        break
      end
      Chef::Log.debug("service[jenkins] still listening (port #{node[:jenkins][:server][:port]})")
      sleep 1
    end
  end
  action :nothing
end


service "jenkins" do
  supports [ :stop, :start, :restart, :status ]
  status_command "test -f #{pid_file} && kill -0 `cat #{pid_file}`"
  action :nothing
end

ruby_block "block_until_operational" do
  block do
    until IO.popen("netstat -lnt").entries.select { |entry|
        entry.split[3] =~ /:#{node[:jenkins][:server][:port]}$/
      }.size == 1
      Chef::Log.debug "service[jenkins] not listening on port #{node.jenkins.server.port}"
      sleep 1
    end

    loop do
      url = URI.parse("#{node.jenkins.server.url}/job/test/config.xml")
      res = Chef::REST::RESTRequest.new(:GET, url, nil).call
      break if res.kind_of?(Net::HTTPSuccess) or res.kind_of?(Net::HTTPNotFound)
      Chef::Log.debug "service[jenkins] not responding OK to GET / #{res.inspect}"
      sleep 1
    end
  end
  action :nothing
end

log "jenkins: install and start" do
  notifies :install, "package[jenkins]", :immediately
  notifies :start, "service[jenkins]", :immediately unless install_starts_service
  notifies :create, "ruby_block[block_until_operational]", :immediately
  not_if do
    File.exists? "/usr/share/jenkins/jenkins.war"
  end
end

bound_interface = node[:jenkins][:server][:url]
case node[:jenkins][:http_proxy][:variant]
when "nginx","apache2"
    bound_interface = "localhost"
end
node[:jenkins][:server][:url]  = "http://#{bound_interface}:#{node[:jenkins][:server][:port]}"

template "/etc/default/jenkins" do
  source "jenkins.erb"
  owner       'root'
  group       'root'
  mode        '0644'
  variables :bound_interface => bound_interface
  notifies  :restart, 'service[jenkins]'
end

template "/etc/init/jenkins.conf" do
  source      "jenkins.conf.erb"
  owner       'root'
  group       'root'
  mode        '0644'
  variables(
    :port => node[:jenkins][:server][:port],
    :java_home => node[:java][:java_home]
  )

  if File.exists?("#{node[:nginx][:dir]}/sites-enabled/jenkins.conf")
    notifies  :restart, 'service[jenkins]'
  end
end

template "#{node[:jenkins][:server][:home]}/hudson.tasks.Maven.xml" do
  owner       'jenkins'
  group       'jenkins'
  mode        '0644'
  variables(
    :m2label => "maven#{node[:maven][:version]}",
    :m2home => node[:maven][:m2_home]
  )
end

package "jenkins" do
  action :nothing
  notifies :create, "template[/etc/default/jenkins]", :immediately
  notifies :create, "template[/etc/init/jenkins.conf]", :immediately
  notifies :create, "template[#{node[:jenkins][:server][:home]}/hudson.tasks.Maven.xml]", :immediately
end

# restart if this run only added new plugins
log "plugins updated, restarting jenkins" do
  #ugh :restart does not work, need to sleep after stop.
  notifies :stop, "service[jenkins]", :immediately
  notifies :create, "ruby_block[netstat]", :immediately
  notifies :start, "service[jenkins]", :immediately
  notifies :create, "ruby_block[block_until_operational]", :immediately
  only_if do
    if File.exists?(pid_file)
      htime = File.mtime(pid_file)
      Dir["#{node[:jenkins][:server][:home]}/plugins/*.hpi"].select { |file|
        File.mtime(file) > htime
      }.size > 0
    end
  end

  action :nothing
end

# Front Jenkins with an HTTP server
case node[:jenkins][:http_proxy][:variant]
when "nginx"
  include_recipe "jenkins::proxy_nginx"
when "apache2"
  include_recipe "jenkins::proxy_apache2"
end

if node.jenkins.iptables_allow == "enable"
  include_recipe "iptables"
  iptables_rule "port_jenkins" do
    if node[:jenkins][:iptables_allow] == "enable"
      enable true
    else
      enable false
    end
  end
end


builds = data_bag('builds')
builds.each do | b |
    build = data_bag_item('builds', b)
    branch = build['branch']
    component = build['component']
    repository = build['repository']
    build_on_change = build['build_on_change']

    job_name = "#{component}-#{branch}"

    job_config = File.join(node[:jenkins][:server][:home], "#{job_name}-config.xml")

    jenkins_job job_name do
      action :nothing
      config job_config
    end


    template job_config do
      owner       'jenkins'
      group       'jenkins'
      mode        '0644'
      source "#{component}-#{branch}-config.xml.erb"
      variables(
        :job_name => job_name, :branch => branch, :node => node[:fqdn], :repository => repository,
        :groupId => build['groupId'],
        :artifactId => build['artifactId'],
        :roles => build['roles'],
        :goal => build['goal'],
        :ami => build['ami'],
        :server_size => build['server_size'],
        :ec2_ssh_key => build['ec2_ssh_key'],
        :ec2_security_group => build['ec2_security_group'],
        :chef_bootstrap => build['chef_bootstrap'],
        :ami_user => build['ami_user'],
        :ec2_ssh_key_file => build['ec2_ssh_key_file'],
        :ec2_region => build['ec2_region'] )
      notifies :update, resources(:jenkins_job => job_name), :immediately
      notifies :build, resources(:jenkins_job => job_name), :delayed if build_on_change
    end

end

